horovod/common/controller.cc

```diff
@@ -0,0 +1,735 @@
+//      as from its own TensorFlow ops, and stores them in a request table.
+//      The coordinator continues to receive Request messages until it has
+//      received GLOBAL_SIZE number of empty "DONE" messages.
+//
+//      d) The coordinator finds all tensors that are ready to be reduced,
+//      gathered, or all operations that result in an error. For each of
+//      those, it sends a Response to all the workers. When no more
+//      Responses are available, it sends a "DONE" response to the workers.
+//      If the process is being shutdown, it instead sends a "SHUTDOWN"
+//      response.
+//
+//      e) The workers listen for Response messages, processing each one by
+//      doing the required reduce or gather, until they receive a "DONE"
+//      response from the coordinator. At that point, the tick ends.
+//      If instead of "DONE" they receive "SHUTDOWN", they mark it in the
+//      response list.
+ResponseList Controller::ComputeResponseList() {
+  // Update cache capacity if autotuning is active.
+  if (parameter_manager_.IsAutoTuning()) {
+    response_cache_.set_capacity((int)parameter_manager_.CacheEnabled() *
+                                 cache_capacity_);
+  }
+
+  // Copy the data structures out from parameters.
+  // However, don't keep the lock for the rest of the loop, so that
+  // enqueued stream callbacks can continue.
+
+  CacheCoordinator cache_coordinator(response_cache_.num_active_bits());
+
+  // message queue used only in this cycle
+  std::deque<Request> message_queue_tmp;
+  tensor_queue_.PopMessagesFromQueue(message_queue_tmp);
+  for (auto& message : message_queue_tmp) {
+    // Keep track of cache hits
+    if (response_cache_.capacity() > 0) {
+      auto cache_ = response_cache_.cached(message);
+      if (cache_ == ResponseCache::CacheState::HIT) {
+        uint32_t cache_bit = response_cache_.peek_cache_bit(message);
+        cache_coordinator.record_hit(cache_bit);
+
+        // Record initial time cached tensor is encountered in queue.
+        stall_inspector_.RecordCacheTensorStart(message.tensor_name());
+
+      } else {
+        if (cache_ == ResponseCache::CacheState::INVALID) {
+          uint32_t cache_bit = response_cache_.peek_cache_bit(message);
+          cache_coordinator.record_invalid_bit(cache_bit);
+        }
+        cache_coordinator.set_uncached_in_queue(true);
+
+        // Remove timing entry if uncached or marked invalid.
+        stall_inspector_.RemoveCachedTensor(message.tensor_name());
+      }
+    }
+  }
+
+  // Flag indicating that the background thread should shut down.
+  bool should_shut_down = shut_down_;
+
+  // Check for stalled tensors.
+  if (stall_inspector_.ShouldPerformCheck()) {
+    if (is_coordinator_) {
+      should_shut_down |= stall_inspector_.CheckForStalledTensors(size_);
+    }
+
+    if (response_cache_.capacity() > 0) {
+      stall_inspector_.InvalidateStalledCachedTensors(cache_coordinator,
```

Comment:
Inject Horovod size, cache_coordinator and response_cache during stall inspector construction?

Reply:
`cache_coordinator` will change across multiple functions calls; `Horovod size` is not yet initialized at construction time. So `response_cache` is the only thing that can be injected.

Reply:
Keep forgetting about cache coordinator.